### PR TITLE
Simplify a few diagonal-related functions

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87182
* #87183
* #87205
* #87204
* #87203
* #87424
* #87181
* #87282
* __->__ #87180
* #87278
* #87202

`diag` was unnecessarily implemented as a kernel rather than as a composite
function, which made it unnecessarily difficult (explicit backward + all it entails).

We also change a few uses of `diag` on 2D tensors for `diagonal()`. The
latter returns a view rather than creating a new tensor.

We also upgrade its meta implementation to a fully-fledged
decomposition

I tried implementing the backwards of `diagonal()` via `diag_scatter` (or better `diag_scatter_` to keep the perf) but functionalisation was failing and I was not sure how to fix this, so I moved on. It may be possible to simplify that one as well if @soulitzer or someone knows how to do this.

### CODE REVIEW

**Comment by lezcano** [comment]
@albanD @soulitzer you have been tagged as codeowners, but regardless, could any of you have a look at this one?

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -3390,72 +3390,31 @@ Tensor unfold(const Tensor& self, int64_t d, int64_t size, int64_t step) {
...
+Tensor& diag_out(const Tensor& self, int64_t offset, Tensor& result) {
+  auto ndim = self.dim();
+  TORCH_CHECK(ndim == 1 || ndim == 2, "Supports 1D or 2D tensors. Got ", self.dim(), "D");
+  TORCH_CHECK(canCast(self.scalar_type(), result.scalar_type()),
+              self.scalar_type(),
+              " can't be cast to the desired output type ", result.scalar_type());
+  if (ndim == 1) {
+    return at::diag_embed_out(result, self, offset);
+  } else {
+    auto d = at::diagonal(self, offset);
```
Comment (albanD) [inline aten/src/ATen/native/TensorShape.cpp:3413]: Should we use `diag_copy()` or `diagonal_copy()` here?

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -3434,6 +3435,22 @@ def vsplit(
...
+    offset: int = 0,
+) -> TensorLikeType:
+    ndim = self.dim()
+    utils.check(
+        ndim in (1, 2), lambda: f"diag(): Supports 1D or 2D tensors. Got {ndim}D"
+    )
+    if ndim == 1:
+        return torch.diag_embed(self, offset)
+    else:
+        return torch.diagonal(self, offset).clone()
```
Comment (albanD) [inline torch/_refs/__init__.py:3451]: diag_copy()

**Review by albanD — APPROVED** [review]
Sounds good. Any perf becnhmark to make sure we don't regress anything too badly?

**Review by mruberry — APPROVED** [review]
Looks good but at least two of the test failures are real

File: aten/src/ATen/native/TensorShape.cpp (hunk 1 above)
Comment (lezcano) [reply aten/src/ATen/native/TensorShape.cpp:3413]: Yup! `diagonal_copy`!

**Comment by lezcano** [comment]
@albanD @bdhirsh this new implementation complains when run on LTC on 1D tensors, as it goes through `diag_embed` and this one has a `copy_` inside. Is there any workaround for this?

**Comment by bdhirsh** [comment]
@lezcano do you have a link to the failure? I looked through the CI failures but I only see test failures in vmap (not in LTC).

For LTC - if the op is `CompositeImplicitAutograd`, then it should just work (functionalization will run underneath the decomposition and turn `copy_()` to `copy()`)

If the op is `CompositeExplicitAutograd`, then some extra stuff is needed for LTC. Functionalization has already ran by the time we hit a `CompositeExplicitAutograd` decomp, so we need to manually "re-functionalize" it. You can do that by:

(1) changing the dispatch table so that the entry is `CompositeExplicitAutogradNonFunctional` (this is mostly so we get better errors)
(2) Follow the pattern for re-functionalizing here. We do something similar for a handful of other ops, like `block_diag`. ([yaml entry](https://github.com/pytorch/pytorch/blob/b3071e2eb61fbbad9b36d7022855111efc6c37f4/aten/src/ATen/native/ts_native_functions.yaml#L191)) and ([LTC kernel entry](https://github.com/pytorch/pytorch/blob/master/torch/csrc/lazy/ts_backend/ts_native_functions.cpp#L432)).

**Comment by lezcano** [comment]
This op is indeed ExplicitAutograd, so I'll try to implement your fix. Thank you!

**Pull request closed without merging**